slashing/program/src/duplicate_block_proof.rs

```diff
@@ -0,0 +1,830 @@
+        shred::{Shred, ShredType},
+        state::{ProofType, SlashingProofData},
+    },
+    serde_derive::{Deserialize, Serialize},
+    solana_program::{clock::Slot, msg, pubkey::Pubkey},
+};
+
+#[derive(Deserialize, Serialize)]
+pub struct DuplicateBlockProofData {
+    #[serde(with = "serde_bytes")]
+    pub shred1: Vec<u8>,
+    #[serde(with = "serde_bytes")]
+    pub shred2: Vec<u8>,
+}
+
+impl SlashingProofData for DuplicateBlockProofData {
+    const PROOF_TYPE: ProofType = ProofType::DuplicateBlockProof;
+
+    fn verify_proof(self, slot: Slot, node_pubkey: Pubkey) -> Result<(), SlashingError> {
+        msg!(
+            "Verifying duplicate proof for {} slot {}",
+            node_pubkey,
+            slot
+        );
+        // TODO: verify through instruction inspection that the shreds were sigverified
+        // earlier in this transaction.
+        // Ed25519 Singature verification is performed on the merkle root:
+        // node_pubkey.verify_strict(merkle_root, signature).
+        // We will verify that the pubkey merkle root and signature match the shred and
+        // that the verification was successful.
+        let shred1 = Shred::new_from_payload(self.shred1)?;
+        let shred2 = Shred::new_from_payload(self.shred2)?;
+        check_shreds(slot, &shred1, &shred2)
+    }
+}
+
+/// Check that `shred1` and `shred2` indicate a valid duplicate proof
+///     - Must be for the same slot `slot`
+///     - Must be for the same shred version
+///     - Must have a merkle root conflict, otherwise `shred1` and `shred2` must
+///       have the same `shred_type`
+///     - If `shred1` and `shred2` share the same index they must be not have
+///       equal payloads excluding the retransmitter signature
+///     - If `shred1` and `shred2` do not share the same index and are data
+///       shreds verify that they indicate an index conflict. One of them must
+///       be the LAST_SHRED_IN_SLOT, however the other shred must have a higher
+///       index.
+///     - If `shred1` and `shred2` do not share the same index and are coding
+///       shreds verify that they have conflicting erasure metas
+fn check_shreds(slot: Slot, shred1: &Shred, shred2: &Shred) -> Result<(), SlashingError> {
+    if shred1.slot()? != slot {
+        msg!(
+            "Invalid proof for different slots {} vs {}",
+            shred1.slot()?,
+            slot,
+        );
+        return Err(SlashingError::SlotMismatch);
+    }
+
+    if shred2.slot()? != slot {
+        msg!(
+            "Invalid proof for different slots {} vs {}",
+            shred1.slot()?,
+            slot,
+        );
+        return Err(SlashingError::SlotMismatch);
+    }
+
+    if shred1.version()? != shred2.version()? {
+        msg!(
+            "Invalid proof for different shred versions {} vs {}",
+            shred1.version()?,
+            shred2.version()?,
+        );
+        return Err(SlashingError::InvalidShredVersion);
+    }
+
```

Comment:
The other difference from gossip is that we do not verify that the `node_pubkey` was supposed to be the leader for `slot` via the leader schedule.

This is important on the gossip side because without this verification anyone could censor a leader in fork choice by submitting a duplicate shred proof for the leader's slot. However for slashing since signing these shreds will get you slashed I believe it is not necessary to incur the overhead of checking the leader schedule.

Reply:
Mm yeah, that all makes sense. It might be worth making that clear in the SIMD.